File: lib/js/__tests__/Aeson_spec_test.js
'use strict';

var Fs = require("fs");
var Js_json = require("bs-platform/lib/js/js_json.js");
var AesonSpec = require("../src/AesonSpec.js");
var Caml_array = require("bs-platform/lib/js/caml_array.js");
var Aeson_decode = require("bs-aeson/lib/js/src/Aeson_decode.js");
var Aeson_encode = require("bs-aeson/lib/js/src/Aeson_encode.js");
var AesonSpec_Jest = require("../src/AesonSpec_Jest.js");
var Caml_js_exceptions = require("bs-platform/lib/js/caml_js_exceptions.js");

function encodePerson(p) {
  return Aeson_encode.object_({
              hd: [
                "name",
                p.name
              ],
              tl: {
                hd: [
                  "age",
                  p.age
                ],
                tl: /* [] */0
              }
            });
}

function brokenEncodePerson(p) {
  return Aeson_encode.object_({
              hd: [
                "Name",
                p.name
              ],
              tl: {
                hd: [
                  "Age",
                  p.age
                ],
                tl: /* [] */0
              }
            });
}

function decodePerson(json) {
  var v;
  try {
    v = {
      name: Aeson_decode.field("name", Aeson_decode.string, json),
      age: Aeson_decode.field("age", Aeson_decode.$$int, json)
    };
  }
  catch (raw_message){
    var message = Caml_js_exceptions.internalToOCamlException(raw_message);
    if (message.RE_EXN_ID === Aeson_decode.DecodeError) {
      return {
              TAG: /* Error */1,
              _0: "decodePerson: " + message._1
            };
    }
    throw message;
  }
  return {
          TAG: /* Ok */0,
          _0: v
        };
}

function brokenDecodePerson(json) {
  var v;
  try {
    v = {
      name: Aeson_decode.field("name!!!!", Aeson_decode.string, json),
      age: Aeson_decode.field("age!!!!!", Aeson_decode.$$int, json)
    };
  }
  catch (raw_message){
    var message = Caml_js_exceptions.internalToOCamlException(raw_message);
    if (message.RE_EXN_ID === Aeson_decode.DecodeError) {
      return {
              TAG: /* Error */1,
              _0: "decodePerson: " + message._1
            };
    }
    throw message;
  }
  return {
          TAG: /* Ok */0,
          _0: v
        };
}

function encodeCompany(p) {
  return Aeson_encode.object_({
              hd: [
                "companyName",
                p.companyName
              ],
              tl: {
                hd: [
                  "employees",
                  Aeson_encode.list(encodePerson, p.employees)
                ],
                tl: /* [] */0
              }
            });
}

function decodeCompany(json) {
  var v;
  try {
    v = {
      companyName: Aeson_decode.field("companyName", Aeson_decode.string, json),
      employees: Aeson_decode.field("employees", (function (param) {
              return Aeson_decode.list((function (a) {
                            return Aeson_decode.unwrapResult(decodePerson(a));
                          }), param);
            }), json)
    };
  }
  catch (raw_message){
    var message = Caml_js_exceptions.internalToOCamlException(raw_message);
    if (message.RE_EXN_ID === Aeson_decode.DecodeError) {
      return {
              TAG: /* Error */1,
              _0: "decodePerson: " + message._1
            };
    }
    throw message;
  }
  return {
          TAG: /* Ok */0,
          _0: v
        };
}

function encodeShape(x) {
  switch (x.TAG | 0) {
    case /* Square */0 :
        return Aeson_encode.object_({
                    hd: [
                      "tag",
                      "Square"
                    ],
                    tl: {
                      hd: [
                        "contents",
                        [
                          x._0,
                          x._1
                        ]
                      ],
                      tl: /* [] */0
                    }
                  });
    case /* Triangle */1 :
        return Aeson_encode.object_({
                    hd: [
                      "tag",
                      "Triangle"
                    ],
                    tl: {
                      hd: [
                        "contents",
                        [
                          x._0,
                          x._1,
                          x._2
                        ]
                      ],
                      tl: /* [] */0
                    }
                  });
    case /* Rectangle */2 :
        return Aeson_encode.object_({
                    hd: [
                      "tag",
                      "Rectangle"
                    ],
                    tl: {
                      hd: [
                        "contents",
                        [
                          x._0,
                          x._1,
                          x._2,
                          x._3
                        ]
                      ],
                      tl: /* [] */0
                    }
                  });
    
  }
}

function decodeShape(json) {
  var exit = 0;
  var err;
  try {
    err = Aeson_decode.field("tag", Aeson_decode.string, json);
    exit = 1;
  }
  catch (raw_message){
    var message = Caml_js_exceptions.internalToOCamlException(raw_message);
    if (message.RE_EXN_ID === Aeson_decode.DecodeError) {
      return {
              TAG: /* Error */1,
              _0: message._1
            };
    }
    throw message;
  }
  if (exit === 1) {
    switch (err) {
      case "Rectangle" :
          var v = Aeson_decode.field("contents", Js_json.decodeArray, json);
          if (v === undefined) {
            return {
                    TAG: /* Error */1,
                    _0: "Rectangle expected an array."
                  };
          }
          var exit$1 = 0;
          var v0;
          try {
            v0 = Aeson_decode.$$int(Caml_array.get(v, 0));
            exit$1 = 2;
          }
          catch (raw_message$1){
            var message$1 = Caml_js_exceptions.internalToOCamlException(raw_message$1);
            if (message$1.RE_EXN_ID === Aeson_decode.DecodeError) {
              return {
                      TAG: /* Error */1,
                      _0: "Rectangle: " + message$1._1
                    };
            }
            throw message$1;
          }
          if (exit$1 === 2) {
            var exit$2 = 0;
            var v1;
            try {
              v1 = Aeson_decode.$$int(Caml_array.get(v, 1));
              exit$2 = 3;
            }
            catch (raw_message$2){
              var message$2 = Caml_js_exceptions.internalToOCamlException(raw_message$2);
              if (message$2.RE_EXN_ID === Aeson_decode.DecodeError) {
                return {
                        TAG: /* Error */1,
                        _0: "Rectangle: " + message$2._1
                      };
              }
              throw message$2;
            }
            if (exit$2 === 3) {
              var exit$3 = 0;
              var v2;
              try {
                v2 = Aeson_decode.$$int(Caml_array.get(v, 2));
                exit$3 = 4;
              }
              catch (raw_message$3){
                var message$3 = Caml_js_exceptions.internalToOCamlException(raw_message$3);
                if (message$3.RE_EXN_ID === Aeson_decode.DecodeError) {
                  return {
                          TAG: /* Error */1,
                          _0: "Rectangle: " + message$3._1
                        };
                }
                throw message$3;
              }
              if (exit$3 === 4) {
                var exit$4 = 0;
                var v3;
                try {
                  v3 = Aeson_decode.$$int(Caml_array.get(v, 3));
                  exit$4 = 5;
                }
                catch (raw_message$4){
                  var message$4 = Caml_js_exceptions.internalToOCamlException(raw_message$4);
                  if (message$4.RE_EXN_ID === Aeson_decode.DecodeError) {
                    return {
                            TAG: /* Error */1,
                            _0: "Rectangle: " + message$4._1
                          };
                  }
                  throw message$4;
                }
                if (exit$4 === 5) {
                  return {
                          TAG: /* Ok */0,
                          _0: {
                            TAG: /* Rectangle */2,
                            _0: v0,
                            _1: v1,
                            _2: v2,
                            _3: v3
                          }
                        };
                }
                
              }
              
            }
            
          }
          break;
      case "Square" :
          var v$1 = Aeson_decode.field("contents", Js_json.decodeArray, json);
          if (v$1 === undefined) {
            return {
                    TAG: /* Error */1,
                    _0: "Square expected an array."
                  };
          }
          var exit$5 = 0;
          var v0$1;
          try {
            v0$1 = Aeson_decode.$$int(Caml_array.get(v$1, 0));
            exit$5 = 2;
          }
          catch (raw_message$5){
            var message$5 = Caml_js_exceptions.internalToOCamlException(raw_message$5);
            if (message$5.RE_EXN_ID === Aeson_decode.DecodeError) {
              return {
                      TAG: /* Error */1,
                      _0: "Square: " + message$5._1
                    };
            }
            throw message$5;
          }
          if (exit$5 === 2) {
            var exit$6 = 0;
            var v1$1;
            try {
              v1$1 = Aeson_decode.$$int(Caml_array.get(v$1, 1));
              exit$6 = 3;
            }
            catch (raw_message$6){
              var message$6 = Caml_js_exceptions.internalToOCamlException(raw_message$6);
              if (message$6.RE_EXN_ID === Aeson_decode.DecodeError) {
                return {
                        TAG: /* Error */1,
                        _0: "Square: " + message$6._1
                      };
              }
              throw message$6;
            }
            if (exit$6 === 3) {
              return {
                      TAG: /* Ok */0,
                      _0: {
                        TAG: /* Square */0,
                        _0: v0$1,
                        _1: v1$1
                      }
                    };
            }
            
          }
          break;
      case "Triangle" :
          var v$2 = Aeson_decode.field("contents", Js_json.decodeArray, json);
          if (v$2 === undefined) {
            return {
                    TAG: /* Error */1,
                    _0: "Triangle expected an array."
                  };
          }
          var exit$7 = 0;
          var v0$2;
          try {
            v0$2 = Aeson_decode.$$int(Caml_array.get(v$2, 0));
            exit$7 = 2;
          }
          catch (raw_message$7){
            var message$7 = Caml_js_exceptions.internalToOCamlException(raw_message$7);
            if (message$7.RE_EXN_ID === Aeson_decode.DecodeError) {
              return {
                      TAG: /* Error */1,
                      _0: "Triangle: " + message$7._1
                    };
            }
            throw message$7;
          }
          if (exit$7 === 2) {
            var exit$8 = 0;
            var v1$2;
            try {
              v1$2 = Aeson_decode.$$int(Caml_array.get(v$2, 1));
              exit$8 = 3;
            }
            catch (raw_message$8){
              var message$8 = Caml_js_exceptions.internalToOCamlException(raw_message$8);
              if (message$8.RE_EXN_ID === Aeson_decode.DecodeError) {
                return {
                        TAG: /* Error */1,
                        _0: "Triangle: " + message$8._1
                      };
              }
              throw message$8;
            }
            if (exit$8 === 3) {
              var exit$9 = 0;
              var v2$1;
              try {
                v2$1 = Aeson_decode.$$int(Caml_array.get(v$2, 2));
                exit$9 = 4;
              }
              catch (raw_message$9){
                var message$9 = Caml_js_exceptions.internalToOCamlException(raw_message$9);
                if (message$9.RE_EXN_ID === Aeson_decode.DecodeError) {
                  return {
                          TAG: /* Error */1,
                          _0: "Triangle: " + message$9._1
                        };
                }
                throw message$9;
              }
              if (exit$9 === 4) {
                return {
                        TAG: /* Ok */0,
                        _0: {
                          TAG: /* Triangle */1,
                          _0: v0$2,
                          _1: v1$2,
                          _2: v2$1
                        }
                      };
              }
              
            }
            
          }
          break;
      default:
        return {
                TAG: /* Error */1,
                _0: "Unknown tag value found '" + (err + "'.")
              };
    }
  }
  
}

var Test = {
  encodePerson: encodePerson,
  brokenEncodePerson: brokenEncodePerson,
  decodePerson: decodePerson,
  brokenDecodePerson: brokenDecodePerson,
  encodeCompany: encodeCompany,
  decodeCompany: decodeCompany,
  encodeShape: encodeShape,
  decodeShape: decodeShape
};

AesonSpec.goldenSpec(decodePerson, encodePerson, "person", "__tests__/golden/Person.json");

AesonSpec.goldenDirSpec(decodePerson, encodePerson, "person", "__tests__/golden/Person");

AesonSpec.goldenDirSpec(decodeCompany, encodeCompany, "company", "__tests__/golden/Company");

AesonSpec.goldenDirSpec(decodeShape, encodeShape, "shape", "__tests__/golden/Shape");

AesonSpec_Jest.describe("isJsonFile", (function (param) {
        return AesonSpec_Jest.test("", (function (param) {
                      var files = Fs.readdirSync("__tests__/golden/Person");
                      var filename = Caml_array.get(files, 0);
                      return AesonSpec_Jest.Expect.toEqual(true, AesonSpec_Jest.Expect.expect(AesonSpec.isJsonFile(filename)));
                    }));
      }));

exports.Test = Test;
/*  Not a pure module */

File: __tests__/Aeson_spec_test.ml
open Jest
open Expect

module Test = struct
  type person = { name : string ; age : int }

  type company = { companyName : string ; employees : person list }

  type shape =
    | Square of int * int
    | Triangle of int * int * int
    | Rectangle of int * int * int * int

  let encodePerson (p : person) :Js_json.t =
    Aeson.Encode.object_
      [ ( "name", Aeson.Encode.string p.name)
      ; ( "age", Aeson.Encode.int p.age)
      ]

  let brokenEncodePerson (p : person) :Js_json.t =
    Aeson.Encode.object_
      [ ( "Name", Aeson.Encode.string p.name)
      ; ( "Age", Aeson.Encode.int p.age)
      ]

  let decodePerson (json : Js_json.t) :(person, string) Belt.Result.t =
    match Aeson.Decode.
          { name = field "name" string json
          ; age = field "age" int json
          }
    with
    | v -> Belt.Result.Ok v
    | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("decodePerson: " ^ message)

  let encodeCompany (p : company) :Js_json.t =
    Aeson.Encode.object_
      [ ( "companyName", Aeson.Encode.string p.companyName)
      ; ( "employees", Aeson.Encode.list encodePerson p.employees)
      ]

  let decodeCompany (json : Js_json.t) :(company, string) Belt.Result.t =
    match Aeson.Decode.
          { companyName = field "companyName" string json
          ; employees = field "employees" (list (fun a -> unwrapResult (decodePerson a))) json
          }
    with
    | v -> Belt.Result.Ok v
    | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("decodePerson: " ^ message)

  let encodeShape (x : shape) :Js_json.t =
    match x with
    | Square (y0,y1) ->
       Aeson.Encode.object_
         [ ( "tag", Aeson.Encode.string "Square" )
         ; ( "contents" , Aeson.Encode.array [| Aeson.Encode.int y0 ; Aeson.Encode.int y1 |] )
         ]
    | Triangle (y0,y1,y2) ->
       Aeson.Encode.object_
         [ ( "tag", Aeson.Encode.string "Triangle" )
         ; ( "contents" , Aeson.Encode.array [| Aeson.Encode.int y0 ; Aeson.Encode.int y1 ; Aeson.Encode.int y2|])
         ]
    | Rectangle (y0,y1,y2,y3) ->
       Aeson.Encode.object_
         [ ( "tag", Aeson.Encode.string "Rectangle" )
         ; ( "contents" , Aeson.Encode.array [| Aeson.Encode.int y0 ; Aeson.Encode.int y1 ; Aeson.Encode.int y2 ; Aeson.Encode.int y3|])
         ]

  let decodeShape (json : Js_json.t) :(shape, string) Belt.Result.t =
    match Aeson.Decode.(field "tag" string json) with
    | "Square" ->
       (match Aeson.Decode.(field "contents" Js.Json.decodeArray json) with
        | Some v ->
           (match Aeson.Decode.int v.(0) with
            | v0 ->
               (match Aeson.Decode.int v.(1) with
                | v1 ->
                   Belt.Result.Ok (Square (v0, v1))
                | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("Square: " ^ message)
               )
            | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("Square: " ^ message)
           )
        | None -> Belt.Result.Error ("Square expected an array.")
       )
    | "Triangle" ->
       (match Aeson.Decode.(field "contents" Js.Json.decodeArray json) with
        | Some v ->
           (match Aeson.Decode.int v.(0) with
            | v0 ->
               (match Aeson.Decode.int v.(1) with
                | v1 ->
                   (match Aeson.Decode.int v.(2) with
                    | v2 ->
                       Belt.Result.Ok (Triangle (v0, v1, v2))
                    | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("Triangle: " ^ message)
                   )
                | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("Triangle: " ^ message)
               )
            | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("Triangle: " ^ message)
           )
        | None -> Belt.Result.Error ("Triangle expected an array.")
       )
    | "Rectangle" ->
       (match Aeson.Decode.(field "contents" Js.Json.decodeArray json) with
        | Some v ->
           (match Aeson.Decode.int v.(0) with
            | v0 ->
               (match Aeson.Decode.int v.(1) with
                | v1 ->
                   (match Aeson.Decode.int v.(2) with
                    | v2 ->
                       (match Aeson.Decode.int v.(3) with
                        | v3 ->
                           Belt.Result.Ok (Rectangle (v0, v1, v2, v3))
                        | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("Rectangle: " ^ message)
                       )
                    | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("Rectangle: " ^ message)
                   )
                | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("Rectangle: " ^ message)
               )
            | exception Aeson.Decode.DecodeError message -> Belt.Result.Error ("Rectangle: " ^ message)
           )
        | None -> Belt.Result.Error ("Rectangle expected an array.")
       )
    | err -> Belt.Result.Error ("Unknown tag value found '" ^ err ^ "'.")
    | exception Aeson.Decode.DecodeError message -> Belt.Result.Error message

end

let () =
  AesonSpec.goldenSpec Test.decodePerson Test.encodePerson "person" "__tests__/golden/Person.json";
  
  AesonSpec.goldenDirSpec Test.decodePerson Test.encodePerson "person" "__tests__/golden/Person";

  AesonSpec.goldenDirSpec Test.decodeCompany Test.encodeCompany "company" "__tests__/golden/Company";

  AesonSpec.goldenDirSpec Test.decodeShape Test.encodeShape "shape" "__tests__/golden/Shape";

  AesonSpec.goldenDirSpec Test.decodeShape Test.encodeShape "shape" "__tests__/golden/Broken";
  
  describe "isJsonFile" (fun () ->         
    test "" (fun () ->
      let files = Node.Fs.readdirSync "__tests__/golden/Person" in

      let filename = Array.get files 0 in

      expect (AesonSpec.isJsonFile filename) |> toEqual true;
    )
  )
